BParkingNano/plugins/TriggeringMuonProducer.cc

```diff
@@ -0,0 +1,173 @@
+      obj.unpackFilterLabels(iEvent, *triggerBits);
+      obj.unpackPathNames(names);
+
+      bool isTriggerMuon = false;
+      for (unsigned h = 0; h < obj.filterIds().size(); ++h)
+	if(obj.filterIds()[h] == 83){ 
+	  isTriggerMuon = true; 
+	  if(debug) std::cout << "\t   Type IDs:   " << 83;  //83 = muon
+	  break;
+	} 
+
+      if(!isTriggerMuon) continue; 
+      for (unsigned h = 0; h < obj.filterLabels().size(); ++h){
+	std::string filterName = obj.filterLabels()[h];
+	if(filterName.find("hltL3") != std::string::npos  && filterName.find("Park") != std::string::npos){
+	  isTriggerMuon = true;
+	  if(debug) std::cout << "\t   Filters:   " << filterName; 
+	  break;
+	}
+	else{ isTriggerMuon = false; }
+      }
+
+      if(!isTriggerMuon) continue;
+      std::vector<float> localMuon;
+      localMuon.push_back(obj.pt());
+      localMuon.push_back(obj.eta());
+      localMuon.push_back(obj.phi());
+      triggeringMuons.push_back(localMuon);
+      if(debug){ std::cout << "\tTrigger object:  pt " << obj.pt() << ", eta " << obj.eta() << ", phi " << obj.phi() << std::endl;
+	// Print trigger object collection and type
+	std::cout << "\t   Collection: " << obj.collection() << std::endl;
+      }
+    }//trigger objects
+
+    if(debug){
+      std::cout << "\n total n of triggering muons = " << triggeringMuons.size() << std::endl;
+      for(auto ij : triggeringMuons){
+	for(auto il : ij) std::cout << " >>> components (pt, eta, phi) = " << il << std::endl;
+      }
+    }
+
+
+    std::unique_ptr<pat::CompositeCandidateCollection> result( new pat::CompositeCandidateCollection );
+
+
+    //now check for reco muons matched to triggering muons
+    edm::Handle<std::vector<pat::Muon>> muonForTrgHandle;
+    iEvent.getByToken(muonSrc_, muonForTrgHandle);
+
+    for(unsigned int iTrg=0; iTrg<muonForTrgHandle->size(); ++iTrg){
+      const pat::Muon & muon1 = (*muonForTrgHandle)[iTrg];
+
+      if(!(muon1.isLooseMuon() && muon1.isSoftMuon(PV))) continue;
+      TLorentzVector recoMuon(muon1.pt(), muon1.eta(), muon1.phi(), MuonMass_);
+
+      float dRMuonMatching = -1.;
+      int muonMatching_index = -1;
+      for(unsigned int ij=0; ij<triggeringMuons.size(); ++ij){
+
+	TLorentzVector trgMuon((triggeringMuons.at(ij))[0], (triggeringMuons.at(ij))[1], (triggeringMuons.at(ij))[2], MuonMass_);
+	float dR = recoMuon.DeltaR(trgMuon);
+	if((dR < dRMuonMatching || dRMuonMatching == -1) && dR < 0.01){
+	  dRMuonMatching = dR;
+	  muonMatching_index = iTrg;
+	  if(debug) std::cout << " dR = " << dR 
+			      << " reco = " << recoMuon.Pt() << " " << recoMuon.Eta() << " " << recoMuon.Phi() << " " 
+			      << " HLT = " << trgMuon.Pt() << " " << trgMuon.Eta() << " " << trgMuon.Phi()
+			      << std::endl;
+	}
+      }
+
+      //save reco muon 
+      // can add p4 of triggering muon in case
+      if(muonMatching_index != -1){
+	pat::CompositeCandidate recoTriggerMuonCand;
```

Comment:
You can just use a `pat::Muon` and embed the necessary info. Also, is probably faster to first `emplace_back` in the vector and then embed stuff using `vector::back` to access the last item

Reply:
ok